src/Microsoft.Health.Dicom.Functions/Indexing/ReindexOperation.Activity.cs

```diff
@@ -0,0 +1,136 @@
+        [FunctionName(nameof(GetExtendedQueryTagsOfOperationAsync))]
+        public async Task<IReadOnlyList<ExtendedQueryTagStoreEntry>> GetExtendedQueryTagsOfOperationAsync([ActivityTrigger] string operationId, ILogger log)
+        {
+            EnsureArg.IsNotNull(log, nameof(log));
+
+            log.LogInformation("Getting extended query tags of {operationId}", operationId);
+            var entries = await _tagOperationStore.GetEntriesOfOperationAsync(operationId);
+            // only process tags which is on Processing
+            return entries.Where(x => x.Status == TagOperationStatus.Processing).Select(y => y.TagStoreEntry).ToList();
+        }
+
+        /// <summary>
+        /// The activity to get next watermarks of an operation.
+        /// </summary>
+        /// <param name="operationId">The operation id.</param>
+        /// <param name="log">The logger.</param>
+        /// <returns>Next watermarks.</returns>
+        [FunctionName(nameof(GetNextWatermarksOfOperationAsync))]
+        public Task<IReadOnlyList<long>> GetNextWatermarksOfOperationAsync([ActivityTrigger] string operationId, ILogger log)
+        {
+            EnsureArg.IsNotNull(log, nameof(log));
+            log.LogInformation("Getting next watermarks of operation {operationId}", operationId);
+            return _tagOperationStore.GetNextWatermarksOfOperationAsync(operationId, _reindexConfig.BatchSize * _reindexConfig.MaxParallelBatches);
+        }
+
+        /// <summary>
+        ///  The activity to update end watermark of an operation.
+        /// </summary>
+        /// <param name="input">The input.</param>
+        /// <param name="log">The log</param>
+        /// <returns>The task.</returns>
+        [FunctionName(nameof(UpdateEndWatermarkOfOperationAsync))]
+        public Task UpdateEndWatermarkOfOperationAsync([ActivityTrigger] UpdateEndWatermarkOfOperationInput input, ILogger log)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+            EnsureArg.IsNotNull(log, nameof(log));
+
+            log.LogInformation($"Updating end watermark of operation: {input.OperationId}"); ;
+            return _tagOperationStore.UpdateEndWatermarkOfOperationAsync(input.OperationId, input.NextWatermark);
+        }
+
+        /// <summary>
+        /// The activity to reindex  Dicom instance.
+        /// </summary>
+        /// <param name="input">The input</param>
+        /// <param name="logger">The log.</param>
+        /// <returns>The task</returns>
+        [FunctionName(nameof(ReindexInstanceAsync))]
+        public Task ReindexInstanceAsync([ActivityTrigger] ReindexInstanceInput input, ILogger logger)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            logger.LogInformation("Reindexing with {input}", input);
+            return _instanceReindexer.ReindexInstanceAsync(input.TagEntries, input.Watermark);
```

Comment:
How will you be able to re-index a batch of data with a single watermark? You'll need a range, right? Unless you plan for this method to only re-index 1 entry.

Reply:
The activity will accept mutiple watermark to support the Batch later.
To reindex multiple watermarks here, we could either call  `_instanceReindexer.ReindexInstanceAsync` in multi-thread way. Or we can make `_instanceReindexer.ReindexInstanceAsync` to take a bunch of watermarks and process within.

Reply:
Yeah, I worry that a 1:1 mapping for activity to watermark isn't the best way to scale, as a given activity could itself be multi-threaded while the activities themselves scale across a cluster. This is fine for the immediate term, but we'll need to change this before we're done with the first release of this feature.

Reply:
changed the input to watermark array. does it solve problem?